// @po-ui/ng-components/esm2015/lib/components/po-chart/po-chart-container/po-chart-line/po-chart-path/po-chart-path.component.js
import { __decorate } from "tslib";
import { ChangeDetectorRef, Component, Input, ViewChild } from '@angular/core';
import { InputBoolean } from '../../../../../decorators';
import { isIE } from '../../../../../utils/util';
const pathDashoffsetDefaultWidth = 0;
export class PoChartPathComponent {
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
        this.isActive = true;
        this.chartLine = false;
    }
    set color(value) {
        this._color = value.includes('po-color') && this.chartLine ? value.replace('po-color', 'po-border-color') : value;
    }
    get color() {
        return this._color;
    }
    set animate(value) {
        this._animate = value;
        if (!this.animate) {
            this.pathWidth = pathDashoffsetDefaultWidth;
        }
    }
    get animate() {
        return this._animate;
    }
    ngAfterViewInit() {
        if (!isIE()) {
            this.pathAnimation();
        }
    }
    pathAnimation() {
        this.pathWidth = this.chartPath.nativeElement.getTotalLength();
        this.changeDetector.detectChanges();
        setTimeout(this.preventsAnimatingAgain.bind(this), 700);
    }
    // Redefine pathWidth para zero após o período da animação para que a linha se expanda normalmente on resize.
    preventsAnimatingAgain() {
        this.pathWidth = pathDashoffsetDefaultWidth;
    }
}
PoChartPathComponent.decorators = [
    { type: Component, args: [{
                selector: '[po-chart-path]',
                template: "<svg:path #chartPath\r\n  class=\"po-chart-path\"\r\n  [class.po-chart-active-path]=\"isActive\"\r\n  [class.po-chart-line-path]=\"chartLine\"\r\n  [class]=\"color?.includes('po-border-color') || color?.includes('po-color') ? color : ''\"\r\n  [attr.fill]=\"chartLine ? 'none' : color\"\r\n  [attr.stroke-dasharray]=\"pathWidth\"\r\n  [attr.stroke-dashoffset]=\"pathWidth\"\r\n  [attr.stroke]=\"chartLine ? color : ''\"\r\n  [attr.d]=\"coordinates\">\r\n</svg:path> "
            },] }
];
PoChartPathComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
PoChartPathComponent.propDecorators = {
    isActive: [{ type: Input, args: ['p-is-active',] }],
    chartLine: [{ type: Input, args: ['p-chart-line',] }],
    coordinates: [{ type: Input, args: ['p-coordinates',] }],
    chartPath: [{ type: ViewChild, args: ['chartPath', { static: true },] }],
    color: [{ type: Input, args: ['p-color',] }],
    animate: [{ type: Input, args: ['p-animate',] }]
};
__decorate([
    InputBoolean()
], PoChartPathComponent.prototype, "isActive", void 0);
__decorate([
    InputBoolean()
], PoChartPathComponent.prototype, "chartLine", void 0);
