// @po-ui/ng-components/esm2015/lib/components/po-chart/po-chart-container/po-chart-axis/po-chart-axis-label/po-chart-axis-label.component.js
import { Component, Input } from '@angular/core';
export class PoChartAxisLabelComponent {
    constructor() {
        this.alignByTheCorners = false;
    }
    trackBy(index) {
        return index;
    }
}
PoChartAxisLabelComponent.decorators = [
    { type: Component, args: [{
                selector: '[po-chart-axis-label]',
                template: "<svg:g>\r\n  <svg:text *ngFor=\"let item of axisXLabelCoordinates; trackBy: trackBy\"\r\n    class=\"po-chart-axis-x-label\" \r\n    dominant-baseline=\"middle\"\r\n    [attr.x]=\"item.xCoordinate\"\r\n    [attr.y]=\"item.yCoordinate\">\r\n    {{ item.label }}\r\n  </svg:text> \r\n</svg:g>\r\n\r\n<svg:g>\r\n  <svg:text *ngFor=\"let item of axisYLabelCoordinates; let first = first; let last = last; trackBy: trackBy\"\r\n    class=\"po-chart-axis-y-label\" \r\n    [class.po-chart-centered-label]=\"!alignByTheCorners\"\r\n    [attr.x]=\"item.xCoordinate\"\r\n    [attr.y]=\"item.yCoordinate\">\r\n    {{ item.label }}\r\n  </svg:text> \r\n</svg:g>"
            },] }
];
PoChartAxisLabelComponent.ctorParameters = () => [];
PoChartAxisLabelComponent.propDecorators = {
    alignByTheCorners: [{ type: Input, args: ['p-align-by-the-corners',] }],
    axisXLabelCoordinates: [{ type: Input, args: ['p-axis-x-label-coordinates',] }],
    axisYLabelCoordinates: [{ type: Input, args: ['p-axis-y-label-coordinates',] }],
    type: [{ type: Input, args: ['p-type',] }]
};
