// @po-ui/ng-components/esm2015/lib/components/po-toolbar/po-toolbar-base.component.js
import { Input, Directive } from '@angular/core';
import { convertToInt } from '../../utils/util';
/**
 * @description
 *
 * O componente `po-toolbar` é um cabeçalho para o título da aplicação e informações de usuário e notificações quando houver necessidade.
 */
export class PoToolbarBaseComponent {
    constructor() {
        /** Se falso, oculta o ícone de notificações. */
        this.showNotification = true;
    }
    /**
     * @optional
     *
     * @description
     *
     * Número de notificações.
     */
    set notificationNumber(value) {
        this._notificationNumber = convertToInt(value, 0);
    }
    get notificationNumber() {
        return this._notificationNumber;
    }
    get isShowProfile() {
        return !!(this.profile || this.profileActions);
    }
}
PoToolbarBaseComponent.decorators = [
    { type: Directive }
];
PoToolbarBaseComponent.propDecorators = {
    actions: [{ type: Input, args: ['p-actions',] }],
    actionsIcon: [{ type: Input, args: ['p-actions-icon',] }],
    profile: [{ type: Input, args: ['p-profile',] }],
    profileActions: [{ type: Input, args: ['p-profile-actions',] }],
    showNotification: [{ type: Input, args: ['p-show-notification',] }],
    title: [{ type: Input, args: ['p-title',] }],
    notificationActions: [{ type: Input, args: ['p-notification-actions',] }],
    notificationNumber: [{ type: Input, args: ['p-notification-number',] }]
};
