// @po-ui/ng-components/esm2015/lib/components/po-toolbar/po-toolbar.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PoAvatarModule } from './../po-avatar/po-avatar.module';
import { PoIconModule } from './../po-icon/po-icon.module';
import { PoPopupModule } from '../po-popup/po-popup.module';
import { PoToolbarActionsComponent } from './po-toolbar-actions/po-toolbar-actions.component';
import { PoToolbarComponent } from './po-toolbar.component';
import { PoToolbarNotificationComponent } from './po-toolbar-notification/po-toolbar-notification.component';
import { PoToolbarProfileComponent } from './po-toolbar-profile/po-toolbar-profile.component';
/**
 * @description
 *
 * Módulo do componente po-toolbar
 *
 */
export class PoToolbarModule {
}
PoToolbarModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, PoAvatarModule, PoPopupModule, PoIconModule],
                declarations: [
                    PoToolbarActionsComponent,
                    PoToolbarComponent,
                    PoToolbarNotificationComponent,
                    PoToolbarProfileComponent
                ],
                exports: [PoToolbarComponent],
                providers: [],
                schemas: []
            },] }
];
