// @po-ui/ng-components/esm2015/lib/components/po-gauge/po-gauge-svg/po-gauge-pointer/po-gauge-pointer.component.js
import { Component, Input, Renderer2, ViewChild } from '@angular/core';
export class PoGaugePointerComponent {
    constructor(renderer) {
        this.renderer = renderer;
        this.afterViewInit = false;
    }
    set coordinates(value) {
        var _a, _b;
        this._coordinates = value;
        if ((_a = this._coordinates) === null || _a === void 0 ? void 0 : _a.radius) {
            this.calculateRadiusScale(this.coordinates.radius);
        }
        if (((_b = this._coordinates) === null || _b === void 0 ? void 0 : _b.hasOwnProperty('pointerDegrees')) && this.afterViewInit) {
            this.applyPointerRotation(this.coordinates.pointerDegrees);
        }
    }
    get coordinates() {
        return this._coordinates;
    }
    ngAfterViewInit() {
        var _a;
        if ((_a = this._coordinates) === null || _a === void 0 ? void 0 : _a.pointerDegrees) {
            this.applyPointerRotation(this.coordinates.pointerDegrees);
        }
        this.afterViewInit = true;
    }
    calculateRadiusScale(radius) {
        const scale = radius / 20;
        const maxScaleValue = 24;
        this.radiusScale = scale < maxScaleValue ? scale : maxScaleValue;
    }
    applyPointerRotation(degrees) {
        this.renderer.setStyle(this.pointer.nativeElement, 'transformOrigin', `${this.coordinates.radius}px ${this.coordinates.radius}px`);
        this.renderer.setStyle(this.pointer.nativeElement, 'transform', `rotate(${degrees}deg)`);
    }
}
PoGaugePointerComponent.decorators = [
    { type: Component, args: [{
                selector: '[po-gauge-pointer]',
                template: "<svg:g>\r\n  <svg:path #pointer\r\n    class=\"po-gauge-pointer\"\r\n    [attr.d]=\"coordinates?.coordinates\"\r\n  >\r\n  </svg:path> \r\n\r\n  <svg:circle \r\n    class=\"po-gauge-pointer-circle\" \r\n    [attr.cx]=\"coordinates?.radius\" \r\n    [attr.cy]=\"coordinates?.radius\" \r\n    [attr.r]=\"radiusScale\"\r\n  >\r\n  </svg:circle>\r\n</svg:g>"
            },] }
];
PoGaugePointerComponent.ctorParameters = () => [
    { type: Renderer2 }
];
PoGaugePointerComponent.propDecorators = {
    pointer: [{ type: ViewChild, args: ['pointer',] }],
    coordinates: [{ type: Input, args: ['p-coordinates',] }]
};
