// @po-ui/ng-components/esm2015/lib/components/po-chart/po-chart-container/po-chart-line/po-chart-series-point/po-chart-series-point.component.js
import { __decorate } from "tslib";
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2 } from '@angular/core';
import { from, timer } from 'rxjs';
import { concatMap, mapTo, scan, tap } from 'rxjs/operators';
import { isIE } from '../../../../../utils/util';
import { InputBoolean } from '../../../../../decorators';
const RADIUS_DEFAULT_SIZE = 5;
const RADIUS_HOVER_SIZE = 10;
const ANIMATION_DURATION_TIME = 700;
export class PoChartSeriesPointComponent {
    constructor(renderer, elementRef) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.chartLine = false;
        this.pointClick = new EventEmitter();
        this.pointHover = new EventEmitter();
        this.radius = RADIUS_DEFAULT_SIZE;
        this._coordinates = [];
        this.animationState = true;
    }
    set color(value) {
        this.strokeColor = value.includes('po-color') ? value.replace('po-color', 'po-border-color') : value;
        this._color = value;
    }
    get color() {
        return this._color;
    }
    set coordinates(value) {
        this._coordinates = value;
        this.coordinates$ = this.displayPointsWithDelay(this._coordinates);
    }
    get coordinates() {
        return this._coordinates;
    }
    trackBy(index) {
        return index;
    }
    onClick(point) {
        const selectedItem = { label: point.label, data: point.data, category: point.category };
        this.pointClick.emit(selectedItem);
    }
    onMouseEnter(event, point) {
        this.setPointAttribute(event.target, true);
        const selectedItem = { label: point.label, data: point.data, category: point.category };
        this.pointHover.emit(Object.assign({ relativeTo: this.relativeTo }, selectedItem));
    }
    onMouseLeave(event) {
        this.setPointAttribute(event.target, false);
    }
    displayPointsWithDelay(coordinates) {
        if (this.animationState && !isIE()) {
            const animationTimer = ANIMATION_DURATION_TIME / coordinates.length;
            return from(coordinates).pipe(concatMap((item, index) => timer(index === 0 || !this.animate ? 0 : animationTimer).pipe(mapTo(item))), scan((acc, curr) => acc.concat(curr), []), tap(() => (this.animationState = false)));
        }
        else {
            return from([coordinates]);
        }
    }
    setPointAttribute(target, isHover) {
        this.renderer.setAttribute(target, 'r', isHover ? RADIUS_HOVER_SIZE.toString() : RADIUS_DEFAULT_SIZE.toString());
        if (this.color.includes('po-color')) {
            this.renderer.setAttribute(target, 'class', isHover ? `${this.strokeColor} ${this.color}` : `po-chart-line-point po-chart-active-point ${this.strokeColor}`);
        }
        else {
            this.renderer[isHover ? 'setStyle' : 'removeStyle'](target, 'fill', isHover ? this.color : undefined);
        }
    }
}
PoChartSeriesPointComponent.decorators = [
    { type: Component, args: [{
                selector: '[po-chart-series-point]',
                template: "<svg:circle *ngFor=\"let item of coordinates$ | async; trackBy: trackBy\" \r\n  [p-tooltip]=\"item.tooltipLabel\"\r\n  [p-append-in-body]='true'\r\n  [p-display-tooltip]=\"!chartLine && item.isActive\"\r\n  p-tooltip-position=\"top\"\r\n  class=\"po-chart-line-point\"\r\n  [class]=\"strokeColor?.includes('po-border-color') ? strokeColor : ''\"\r\n  [class.po-chart-active-point]=\"item.isActive\"\r\n  [attr.cx]=\"item.xCoordinate\"\r\n  [attr.cy]=\"item.yCoordinate\"\r\n  [attr.r]=\"radius\"\r\n  [attr.stroke]=\"strokeColor\"\r\n  (click)=\"onClick(item)\"\r\n  (mouseenter)=\"onMouseEnter($event, item)\"\r\n  (mouseleave)=\"onMouseLeave($event)\"\r\n  >\r\n</svg:circle>\r\n"
            },] }
];
PoChartSeriesPointComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
PoChartSeriesPointComponent.propDecorators = {
    animate: [{ type: Input, args: ['p-animate',] }],
    isActive: [{ type: Input, args: ['p-is-active',] }],
    chartLine: [{ type: Input, args: ['p-chart-line',] }],
    relativeTo: [{ type: Input, args: ['p-relative-to',] }],
    pointClick: [{ type: Output, args: ['p-point-click',] }],
    pointHover: [{ type: Output, args: ['p-point-hover',] }],
    color: [{ type: Input, args: ['p-color',] }],
    coordinates: [{ type: Input, args: ['p-coordinates',] }]
};
__decorate([
    InputBoolean()
], PoChartSeriesPointComponent.prototype, "isActive", void 0);
__decorate([
    InputBoolean()
], PoChartSeriesPointComponent.prototype, "chartLine", void 0);
