// @po-ui/ng-components/esm2015/lib/components/po-chart/po-chart-container/po-chart-bar/po-chart-bar-path/po-chart-bar-path.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
export class PoChartBarPathComponent {
    constructor() {
        this.barClick = new EventEmitter();
        this.barHover = new EventEmitter();
    }
    trackBy(index) {
        return index;
    }
    onClick(item) {
        const selectedItem = { label: item.label, data: item.data, category: item.category };
        this.barClick.emit(selectedItem);
    }
    onMouseEnter(item) {
        const selectedItem = { label: item.label, data: item.data, category: item.category };
        this.barHover.emit(selectedItem);
    }
}
PoChartBarPathComponent.decorators = [
    { type: Component, args: [{
                selector: '[po-chart-bar-path]',
                template: "<svg:path *ngFor=\"let item of coordinates; trackBy: trackBy\"\r\n  [p-tooltip]=\"item.tooltipLabel\"\r\n  [p-tooltip-position]=\"tooltipPosition\"\r\n  [p-append-in-body]='true'\r\n  class=\"po-chart-bar-path\"\r\n  [class]=\"color?.includes('po-color') ? color : ''\"\r\n  [attr.fill]=\"color\"\r\n  [attr.d]=\"item.coordinates\"\r\n  (mouseenter)=\"onMouseEnter(item)\"\r\n  (click)=\"onClick(item)\">\r\n</svg:path> \r\n"
            },] }
];
PoChartBarPathComponent.ctorParameters = () => [];
PoChartBarPathComponent.propDecorators = {
    color: [{ type: Input, args: ['p-color',] }],
    coordinates: [{ type: Input, args: ['p-coordinates',] }],
    tooltipPosition: [{ type: Input, args: ['p-tooltip-position',] }],
    barClick: [{ type: Output, args: ['p-bar-click',] }],
    barHover: [{ type: Output, args: ['p-bar-hover',] }]
};
