// @po-ui/ng-components/esm2015/lib/components/po-gauge/po-gauge-legend/po-gauge-legend.component.js
import { Component, Input } from '@angular/core';
export class PoGaugeLegendComponent {
    constructor() { }
    set ranges(value) {
        this._ranges = value.length ? this.filterLabel(value) : [];
    }
    get ranges() {
        return this._ranges;
    }
    trackBy(index) {
        return index;
    }
    filterLabel(ranges) {
        return ranges.filter(range => range.label);
    }
}
PoGaugeLegendComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-gauge-legend',
                template: "<ul class=\"po-gauge-legend\">\r\n  <li class=\"po-gauge-legend-item\" *ngFor=\"let range of ranges; trackBy: trackBy\">\r\n    <ng-container\r\n      *ngTemplateOutlet=\"range.color?.includes('po-color') ? colorpalette : colorstyle; context: { $implicit: range }\"\r\n    ></ng-container>\r\n    <p class=\"po-gauge-legend-text\">{{ range.label }}</p>\r\n  </li>\r\n</ul>\r\n\r\n<ng-template #colorpalette let-range>\r\n  <span class=\"po-gauge-legend-square\" [class]=\"range.color\"></span>\r\n</ng-template>\r\n\r\n<ng-template #colorstyle let-range>\r\n  <span class=\"po-gauge-legend-square\" [style.background]=\"range.color\"></span>\r\n</ng-template>\r\n"
            },] }
];
PoGaugeLegendComponent.ctorParameters = () => [];
PoGaugeLegendComponent.propDecorators = {
    ranges: [{ type: Input, args: ['p-ranges',] }]
};
