// @po-ui/ng-components/esm2015/lib/components/po-calendar/po-calendar-header/po-calendar-header.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
export class PoCalendarHeaderComponent {
    constructor() {
        this.hidePrevious = false;
        this.hideNext = false;
        this.previous = new EventEmitter();
        this.next = new EventEmitter();
    }
}
PoCalendarHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-calendar-header',
                template: "<div class=\"po-calendar-header\">\r\n  <span\r\n    *ngIf=\"previous?.observers && !hidePrevious\"\r\n    class=\"po-calendar-header-left po-icon po-icon-arrow-left\"\r\n    (click)=\"previous.emit()\"\r\n  >\r\n  </span>\r\n\r\n  <div class=\"po-calendar-header-title\" attr-calendar>\r\n    <ng-content></ng-content>\r\n  </div>\r\n\r\n  <span\r\n    *ngIf=\"next?.observers.length && !hideNext\"\r\n    class=\"po-calendar-header-right po-icon po-icon-arrow-right\"\r\n    (click)=\"next.emit()\"\r\n  >\r\n  </span>\r\n</div>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
PoCalendarHeaderComponent.ctorParameters = () => [];
PoCalendarHeaderComponent.propDecorators = {
    hidePrevious: [{ type: Input, args: ['p-hide-previous',] }],
    hideNext: [{ type: Input, args: ['p-hide-next',] }],
    previous: [{ type: Output, args: ['p-previous',] }],
    next: [{ type: Output, args: ['p-next',] }]
};
