// @po-ui/ng-components/esm2015/lib/components/po-chart/po-chart-legend/po-chart-legend.component.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
export class PoChartLegendComponent {
    set series(value) {
        this._series = value;
    }
    get series() {
        return this._series;
    }
}
PoChartLegendComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-chart-legend',
                template: "<div class=\"po-chart-legend\">\r\n  <div class=\"po-chart-legend-container\">\r\n    <div class=\"po-chart-legend-item\" *ngFor=\"let serie of series; let i = index\">\r\n      <div\r\n        class=\"po-chart-legend-square\"\r\n        [class]=\"serie.color?.includes('po-color') ? serie.color : ''\"\r\n        [style.background]=\"serie.color\"\r\n      ></div>\r\n      <span class=\"po-chart-legend-text\">{{ serie.label || serie.category }}</span>\r\n    </div>\r\n  </div>\r\n</div>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
PoChartLegendComponent.propDecorators = {
    type: [{ type: Input, args: ['p-type',] }],
    series: [{ type: Input, args: ['p-series',] }]
};
