// @po-ui/ng-components/esm2015/lib/components/po-gauge/po-gauge-description/po-gauge-description.component.js
import { Component, Input, ViewChild } from '@angular/core';
export class PoGaugeDescriptionComponent {
    get isValidValue() {
        return this.value || this.value === 0;
    }
    verifyIfHasEllipsis() {
        const { offsetWidth, scrollWidth } = this.descriptionText.nativeElement;
        // O tooltip só exibe se adicionado dentro de setTimeout.
        setTimeout(() => {
            this.tooltip = offsetWidth < scrollWidth ? this.description : undefined;
        });
    }
}
PoGaugeDescriptionComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-gauge-description',
                template: "<div class=\"po-gauge-description\">\r\n  <ng-container *ngTemplateOutlet=\"hasRanges ? rangesDescription : percentageDescription\"> </ng-container>\r\n</div>\r\n\r\n<ng-template #rangesDescription>\r\n  <p\r\n    #descriptionText\r\n    *ngIf=\"description\"\r\n    class=\"po-gauge-description-text po-gauge-description-upper-text\"\r\n    [p-tooltip]=\"tooltip\"\r\n    (mouseenter)=\"verifyIfHasEllipsis()\"\r\n  >\r\n    {{ description }}\r\n  </p>\r\n</ng-template>\r\n\r\n<ng-template #percentageDescription>\r\n  <div class=\"po-gauge-description-bottom\" tabindex=\"0\" [style.width.px]=\"descriptionWidth\">\r\n    <p *ngIf=\"isValidValue\" class=\"po-gauge-description-bottom-value\">{{ value }}%</p>\r\n    <p\r\n      #descriptionText\r\n      *ngIf=\"description\"\r\n      class=\"po-gauge-description-text po-gauge-description-bottom-text\"\r\n      [p-tooltip]=\"tooltip\"\r\n      (mouseenter)=\"verifyIfHasEllipsis()\"\r\n    >\r\n      {{ description }}\r\n    </p>\r\n  </div>\r\n</ng-template>\r\n"
            },] }
];
PoGaugeDescriptionComponent.propDecorators = {
    description: [{ type: Input, args: ['p-description',] }],
    hasRanges: [{ type: Input, args: ['p-has-ranges',] }],
    descriptionWidth: [{ type: Input, args: ['p-description-width',] }],
    value: [{ type: Input, args: ['p-value',] }],
    descriptionText: [{ type: ViewChild, args: ['descriptionText',] }]
};
