// @po-ui/ng-components/esm2015/lib/components/po-navbar/po-navbar-item-navigation/po-navbar-item-navigation.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
export class PoNavbarItemNavigationComponent {
    constructor() {
        this.click = new EventEmitter();
    }
}
PoNavbarItemNavigationComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-navbar-item-navigation',
                template: "<po-navbar-item-navigation-icon\r\n  class=\"po-navbar-item-navigation-icon\"\r\n  p-icon=\"left\"\r\n  [p-disabled]=\"disableLeft\"\r\n  (p-click)=\"click.emit($event)\"\r\n>\r\n</po-navbar-item-navigation-icon>\r\n\r\n<po-navbar-item-navigation-icon\r\n  class=\"po-navbar-item-navigation-icon\"\r\n  p-icon=\"right\"\r\n  [p-disabled]=\"disableRight\"\r\n  (p-click)=\"click.emit($event)\"\r\n>\r\n</po-navbar-item-navigation-icon>\r\n"
            },] }
];
PoNavbarItemNavigationComponent.propDecorators = {
    disableLeft: [{ type: Input, args: ['p-disable-left',] }],
    disableRight: [{ type: Input, args: ['p-disable-right',] }],
    click: [{ type: Output, args: ['p-click',] }]
};
