// @po-ui/ng-components/esm2015/lib/components/po-toolbar/po-toolbar-profile/po-toolbar-profile.component.js
import { Component, Input } from '@angular/core';
import { PoControlPositionService } from '../../../services/po-control-position/po-control-position.service';
/**
 * @docsPrivate
 *
 * @docsExtends PoToolbarProfileComponent
 *
 * @description
 *
 * Componente `po-toolbar-profile`.
 */
export class PoToolbarProfileComponent {
    get profileAvatar() {
        return this.profile ? this.profile.avatar : undefined;
    }
}
PoToolbarProfileComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-toolbar-profile',
                template: "<div #profileElement class=\"po-toolbar-profile po-clickable\" (click)=\"popup.toggle()\">\r\n  <po-avatar p-size=\"xs\" [p-src]=\"profileAvatar\"></po-avatar>\r\n</div>\r\n\r\n<po-popup #popup [p-actions]=\"profileActions\" [p-target]=\"profileElement\">\r\n  <div p-popup-header-template *ngIf=\"profile?.title\" class=\"po-toolbar-profile-item-header\">\r\n    <po-avatar class=\"po-toolbar-profile-item-avatar\" p-size=\"sm\" [p-src]=\"profileAvatar\"> </po-avatar>\r\n\r\n    <div>\r\n      <div class=\"po-toolbar-profile-item-header-title\">\r\n        {{ profile.title }}\r\n      </div>\r\n\r\n      <div *ngIf=\"profile.subtitle\" class=\"po-toolbar-profile-item-header-subtitle\">\r\n        {{ profile.subtitle }}\r\n      </div>\r\n    </div>\r\n  </div>\r\n</po-popup>\r\n",
                providers: [PoControlPositionService]
            },] }
];
PoToolbarProfileComponent.propDecorators = {
    profile: [{ type: Input, args: ['p-profile',] }],
    profileActions: [{ type: Input, args: ['p-profile-actions',] }]
};
