// @po-ui/ng-components/esm2015/lib/components/po-field/po-multiselect/po-multiselect-item/po-multiselect-item.component.js
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente que constrói cada item no dropdown, contendo o checkbox e o label.
 */
export class PoMultiselectItemComponent {
    constructor() {
        /** Esta propriedade indica se o campo está selecionado, indenterminate ou não. */
        this.selected = false;
        /** Evento que será disparado toda vez que o usuário marcar ou desmarcar um item. */
        this.change = new EventEmitter();
    }
    itemClicked() {
        this.selected = !this.selected;
        this.change.emit(this.selected);
    }
}
PoMultiselectItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-multiselect-item',
                template: "<li [class.po-multiselect-item-selected]=\"selected\" (click)=\"itemClicked()\">\r\n  <a class=\"po-multiselect-item\">\r\n    <input\r\n      [checked]=\"selected\"\r\n      class=\"po-multiselect-checkbox-input\"\r\n      [class.po-multiselect-checkbox-input-indeterminate]=\"selected === null\"\r\n      [class.po-multiselect-checkbox-input-checked]=\"selected\"\r\n      type=\"checkbox\"\r\n    />\r\n\r\n    <label class=\"po-multiselect-checkbox-label po-clickable\">\r\n      {{ label }}\r\n    </label>\r\n  </a>\r\n</li>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
PoMultiselectItemComponent.propDecorators = {
    label: [{ type: Input, args: ['p-label',] }],
    selected: [{ type: Input, args: ['p-selected',] }],
    change: [{ type: Output, args: ['p-change',] }]
};
