// @po-ui/ng-components/esm2015/lib/components/po-slide/po-slide-circles/po-slide-circles.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para o conjunto de círculos que faz a passagem de itens do `po-slide`.
 */
export class PoSlideCirclesComponent {
    constructor() {
        /** Evento emitido ao clicar em um controle. */
        this.click = new EventEmitter();
    }
}
PoSlideCirclesComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-slide-circles',
                template: "<div class=\"po-slide-circles\">\r\n  <ng-container *ngFor=\"let item of items; let index = index\">\r\n    <div\r\n      class=\"po-slide-circle\"\r\n      [class.po-slide-active-circle]=\"index === currentSlideIndex\"\r\n      (click)=\"click.emit(index)\"\r\n    ></div>\r\n  </ng-container>\r\n</div>\r\n"
            },] }
];
PoSlideCirclesComponent.propDecorators = {
    currentSlideIndex: [{ type: Input, args: ['p-current-slide-index',] }],
    items: [{ type: Input, args: ['p-items',] }],
    click: [{ type: Output, args: ['p-click',] }]
};
