// @po-ui/ng-components/esm2015/lib/components/po-chart/po-chart-container/po-chart-circular/po-chart-circular-path/po-chart-circular-path.component.js
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewChild } from '@angular/core';
export class PoChartCircularPathComponent {
    constructor(renderer) {
        this.renderer = renderer;
        this.onClick = new EventEmitter();
        this.onHover = new EventEmitter();
    }
    applyCoordinates(coordinates) {
        this.renderer.setAttribute(this.svgPath.nativeElement, 'd', coordinates);
    }
    onMouseClick() {
        const { label, data } = this.serie;
        this.onClick.emit({ label, data });
    }
    onMouseEnter() {
        const { label, data } = this.serie;
        this.onHover.emit({ label, data });
    }
}
PoChartCircularPathComponent.decorators = [
    { type: Component, args: [{
                selector: '[po-chart-circular-path]',
                template: "<svg:path #svgPath\r\n  class=\"po-path-item\"\r\n  [class]=\"serie?.color.includes('po-color') ? serie.color : ''\"\r\n  [p-chart-tooltip]=\"serie?.tooltipLabel\"\r\n  [attr.fill]=\"serie?.color\"\r\n  (click)=\"onMouseClick()\"\r\n  (mouseenter)=\"onMouseEnter()\"\r\n  >\r\n</svg:path> \r\n"
            },] }
];
PoChartCircularPathComponent.ctorParameters = () => [
    { type: Renderer2 }
];
PoChartCircularPathComponent.propDecorators = {
    serie: [{ type: Input, args: ['p-serie',] }],
    onClick: [{ type: Output, args: ['p-on-click',] }],
    onHover: [{ type: Output, args: ['p-on-hover',] }],
    svgPath: [{ type: ViewChild, args: ['svgPath', { read: ElementRef },] }]
};
